hoomd/mpcd/CollisionMethod.cc

```diff
@@ -48,15 +49,362 @@ void mpcd::CollisionMethod::collide(uint64_t timestep)
     // sync the embedded group
     m_cl->setEmbeddedGroup(m_embed_group);
 
+// check for collision warnings
+#ifdef ENABLE_HIP
+    if (m_exec_conf->isCUDAEnabled())
+        {
+        checkCollisionWarningsGPU(timestep);
+        }
+    else
+#endif
+        {
+        checkCollisionWarnings(timestep);
+        }
+
     // set random grid shift
     m_cl->drawGridShift(timestep);
 
     // update cell list
     m_cl->compute(timestep);
 
+    // create auxillary array for rigid bodies
+    const bool rigid_body_collision = m_embed_group && m_rigid_bodies;
+    if (rigid_body_collision)
+        {
+        // resize initial velocity array
+        const unsigned int num_group = m_embed_group->getNumMembers();
+        if (num_group > m_initial_velocity.getNumElements())
+            {
+            GPUArray<Scalar4> initial_velocity(num_group, m_exec_conf);
+            m_initial_velocity.swap(initial_velocity);
+            }
+
+        const unsigned int num_total = m_pdata->getN();
+        if (num_total > m_linmom_accum.getNumElements())
+            {
+            GPUArray<Scalar3> linmom_accum(num_total, m_exec_conf);
+            m_linmom_accum.swap(linmom_accum);
+            GPUArray<Scalar3> angmom_accum(num_total, m_exec_conf);
+            m_angmom_accum.swap(angmom_accum);
+            }
+#ifdef ENABLE_HIP
+        if (m_exec_conf->isCUDAEnabled())
+            {
+            storeInitialEmbeddedGroupVelocitiesGPU(timestep);
+            }
+        else
+#endif
+            {
+            storeInitialEmbeddedGroupVelocities(timestep);
+            }
+        }
+
+    // apply collisions
     rule(timestep);
+
+    // apply collisions to rigid bodies
+    if (rigid_body_collision)
+        {
+#ifdef ENABLE_HIP
+        if (m_exec_conf->isCUDAEnabled())
+            {
+            accumulateRigidBodyMomentaGPU(timestep);
+            transferRigidBodyMomentaGPU(timestep);
+            m_rigid_bodies->updateCompositeParticles(timestep);
+            }
+        else
+#endif
+            {
+            accumulateRigidBodyMomenta(timestep);
+            transferRigidBodyMomenta(timestep);
+            m_rigid_bodies->updateCompositeParticles(timestep);
+            }
```

Comment:
`updateCompositeParticles` can be called outside the if-else because the type of `ForceComposite` will dispatch this correctly on CPU or GPU.
```suggestion
#ifdef ENABLE_HIP
        if (m_exec_conf->isCUDAEnabled())
            {
            accumulateRigidBodyMomentaGPU(timestep);
            transferRigidBodyMomentaGPU(timestep);
            }
        else
#endif
            {
            accumulateRigidBodyMomenta(timestep);
            transferRigidBodyMomenta(timestep);
            }
            m_rigid_bodies->updateCompositeParticles(timestep);
```